File: rust/src/offer/t24.rs
use std::option::Option::Some;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
            next: None,
            val,
        }
    }
}

pub fn reverse_list(mut head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    let mut h = ListNode::new(0);
    while let Some(mut head_unwrap)=head{
        let tmp = head_unwrap.next.take();
        head_unwrap.next = h.next;
        h.next = Some(head_unwrap);
        head = tmp;
    }
    h.next
}

File: c++/offer/T24.cpp
//
// Created by jesse on 2021/1/2.
//


#include <cstring>

struct ListNode {
    int val;
    ListNode *next;

    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *reverseList(ListNode *head) {
//        if (!head) return NULL;
//        if (head->next==NULL) {
//            return head;
//        }
//        ListNode * last = reverseList(head->next);
//        head->next->next = head;
//        head->next = NULL;
//        return last;
        ListNode *pre = NULL, *curr = head, *next = head;
        while (!curr) {
            next=curr->next;
            curr->next=pre;
            curr=next;
            pre=curr;
        }
        return pre;
    }
};